### ISSUE TITLE
[Question HW1] How to use `finish_job` function?

### ISSUE THREAD

**Issue by DaeyongKwon98** [kaist-cp/cs431#633]
In thread_pool, I pass except `thread_pool_join_block`.
Watching similar questions in issue tracker, there might be problems in
`job count`, `wait_empty`, `join` and `drop` function.

With `job count`, I increment jobcount at `start_job` and decrement at `finish_job`.
I think there's no problem in here.

But I don't know way to make `wait_empty`, `join`, `drop` function.
I think I have to use `ThreadPoolInner::empty_condvar` in `wait_empty` function, maybe.

When I execute my code, I find that my job count increases to 1024.
I put `start_job` at `execute`, but I don't know where to use `finish_job`.

Thank you.

**Comment by Lee-Janggun** [comment]
All I can say is: you should call `finish_job` when the job finishes. If all that you fail is `thread_pool_join_block`, you should be able to add very minimal code to finish it.

Also, please don't post your source code...

**Comment by DaeyongKwon98** [comment]
I think about where to call `finish job`.
`wait_empty` function waits until job count becones 0, so `finish_job` must be called other place.

In my opinion, `finish job` should be called when the job finishes - which means when `thread` is `join`ed. So `finish job` can be inserted among three candidates.

1. `Drop for Worker`
2. `join` function in `impl ThreadPool`
3. `Drop for ThreadPool`

For 1) In `Drop for Worker`, there's no `ThreadPoolInner` in parameter `&mut self`.
So it can't call `finish job` function.

So I tried some cases with rest two options, but in both case, `wait_empty` function goes to infinite loop and never finishes... (I think this is because my `finish job` doesn't work properly, so job count never goes to 0.)

Let me explain my structure(logic).

In `ThreadPool::join`, I just call `wait_empty` function.

```Rust
impl Drop for ThreadPool {
    fn drop(&mut self) {
        drop(self.sender.take());

        for worker in &mut self.workers {
            println!("Shutting down worker {}", worker.id);

            if let Some(thread) = worker.thread.take() {
                thread.join().unwrap();
            }
        }
    }
}
```
This is exact code from Rust book 20.3, and I split this code into `Drop for Worker` and `Drop for ThreadPool`.
I `drop` `sender` in `Drop for ThreadPool`, and use rest of the code into `Drop for Worker`. But I don't use for loop while dropping worker. I think this can be a problem, maybe.

Sorry for asking a poorly organized question.
You said it can be finished with adding very minimal code,
but I don't know if my problem is in `Worker::Drop` or `ThreadPoolInner::wait_empty` or `ThreadPool::join` or `ThreadPool::Drop`.
So I want to start clarifying with `finish job`, not to make infinite loop first.

**Comment by Lee-Janggun** [comment]
I think the point you are misunderstanding is here:
> In my opinion, `finish_job` should be called when the job finishes - which means when `thread` is `join`ed. 

The first part is correct: `finish_job` notifies the `ThreadPoolInner` that this job has finished. 

However the second part is incorrect: the time that the job finishes is not when the thread is `join`ed: it is when the `f` sent via `execute` has completed. 

The purpose of `ThreadPoolInner::join` is for the current thread to **detect** that all jobs have finished, not to **notify** that all jobs have finished.

For example, consider the following situation.

```
time
Thread 1  |   execute(do_long)              join()
Thread 2  |                      do_long() { (* takes a long time *) } <- finish_job() for do_long called
```

Here, since `join()` is called before `do_long()` completed, and `join()` will need to wait until `finish_job()` for `do_long()` is called.

On the other hand, consider the following

```
time
Thread 1  |   execute(do_short)                join()
Thread 2  |                      do_short() <- finish_job() for do_short called
```

Since `join()` is called after `do_short()` completed, it should return immediately, and to do so, `finish_job()` is called right after `do_short()` is done.

I think your main point of confusion was considering which thread's job is to called `finish_job()` is. The 3 places you considered are all the job of the main thread that will spawn the worker threads. 

If that thread were to call `finish_job()`, it will need to detect when the job done by a worker thread has completed and call `finish_job()`. This is very hard, if not impossible, to do without some other synchronization methods. It will be much more simpler if the worker thread calls `finish_job` when the work is done.

**Comment by DaeyongKwon98** [comment]
You said 'The three places you considered gave the role to the main thread that will spawn the worker threads. It will be much more simpler if the worker thread calls finish_job when the work is done.'

In my understanding, my three candidates to use `finish_job` (`Worker::drop`, `ThreadPool::join`, `ThreadPool::drop`) give the role to the main thread, so I need to use `finish_job` in other place.

Then the remaining candidates are `ThreadPool::new` and `ThreadPool::execute`. I can't find where it signals that work is done (where I need to call `finish_job`).

Below is my understanding about each part of the code.

`Worker::drop` join worker thread.
`ThreadPool::join` wait until all threads in threadpool are joined, and
`wait_empty` do exactly what it needs.
`ThreadPool::drop` makes all worker threads to `drop`.

Is there any misconception in my understanding?

Also,
[https://doc.rust-lang.org/std/thread/struct.JoinHandle.html#method.join](url)
In this link, it says `join` waits for the associated thread to finish.
I understand this to mean joining thread means thread finish the job.
If my understanding is correct, I think `finish_job` can be in `Worker::drop`, where I join the worker thread.

Actually, I tried a lot to find the answer myself, but it was not easy.
Sorry for wasting your precious time with my questions. :(

Thank you.

**Comment by Lee-Janggun** [comment]
> Then the remaining candidates are ThreadPool::new and ThreadPool::execute. I can't find where it signals that work is done (where I need to call finish_job).

As I said above: the best thread to call `finish_job` will be the worker thread that actually does the job, and the best place is simply after the call to the job function it is given.

> Below is my understanding about each part of the code.

Yes your understanding is correct. Actually for `drop`, even if you don't do anything, Rust will automatically drop all worker threads, (see [the reference](https://doc.rust-lang.org/reference/destructors.html#destructors))

> I understand this to mean joining thread means thread finish the job.

That is slightly incorrect. A thread T1 calling `join()` on another thread T2 **waits** for  T2 to finish. When T2 is joined, it will have finished. However, even if T2 is not joined, it may have finished.

> If my understanding is correct, I think `finish_job` can be in `Worker::drop`, where I join the worker thread.

This is not correct, as the jobs given to `Worker` may have finished finished before the `Worker` dropped. 
Also,  since the `Worker` may have received many jobs, if you were to call `finish_job` inside `Worker::drop()`, you would need to know how much job the dropping worker had, which is impossible.

**Comment by Lee-Janggun** [comment]
One more point: There is a difference between a **thread** finishing and a **job** finishing. 

A thread finishes when the all functions it needs to run (for a spawned thread, the function given), and a job finishes when the `job` function received by the spawned thread is done. 

`Joinhandle::join()` waits for the former. `ThreadPoolInner::join()` waits for the latter.

**Comment by DaeyongKwon98** [comment]
Finally I'm implementing `wait_empty` function.
From [https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html](url), I tried similar structure with this example.

```rust
let pair = Arc::new((Mutex::new(false), Condvar::new()));
let pair2 = Arc::clone(&pair);

// Inside of our lock, spawn a new thread, and then wait for it to start.
thread::spawn(move|| {
    let (lock, cvar) = &*pair2;
    let mut started = lock.lock().unwrap();
    *started = true;
    // We notify the condvar that the value has changed.
    cvar.notify_one();
});

// Wait for the thread to start up.
let (lock, cvar) = &*pair;
let mut started = lock.lock().unwrap();
while !*started {
    started = cvar.wait(started).unwrap();
}
```
From this, I try to change boolean `*started = true` with condition
`jobcount == 0`, and change `cvar` to `self.empty_condvar`.
Then error says `error[E0521]: borrowed data escapes outside of associated function`. In my opinion, it occurs because I move `condvar` in spawned thread, and after that its lifetime ends.
So I try not to use `self.empty_condvar`, instead use `Condvar::new()` same as example code. Then it goes into infinite loop.

In addition to problem w.r.t `wait_empty`, I'm not sure about my `ThreadPool::drop` implementation. Currently my `ThreadPool::drop` function just `take` and `drop` `sender`.

This is code from rust book.
```rust
impl Drop for ThreadPool {
    fn drop(&mut self) {
        drop(self.sender.take());

        for worker in &mut self.workers {
            println!("Shutting down worker {}", worker.id);

            if let Some(thread) = worker.thread.take() {
                thread.join().unwrap();
            }
        }
    }
}
```
From this code, last if statement goes into `Worker::drop` in my case,
and `take`ing, `drop`ing `sender` remains at `ThreadPool::drop`.
Here I wonder about `for worker in &mut self.workers`.
When threadpool is dropped, all worker threads in pool should be joined.
Can I join all workers in pool without for loop?
If not, my problem with infinite loop may be due to this.

To summarize, I have infinite loop problem, and my problem can be among these.
1. Problem in `wait_empty` function code.
2. `wait_empty` is fine, but `job_count` doesn't go to 0. (In this case, `start_job` and `finish_job`'s position can be problem.)
3. Problem with `drop`, because I omit to implement `for` loop.

Maybe case 2 is not problem, because in `ThreadPool::new` function, it has simple `start_job()` -> `job()` -> `finish_job()` structure.

**Comment by Lee-Janggun** [comment]
> So I try not to use `self.empty_condvar`, instead use `Condvar::new()` same as example code. Then it goes into infinite loop.

Do you mean that you call `Condvar::new()` inside `wait_empty`? This is defiantly weird. What we intended was to use the `empty_condvar` and `job_count` of the `ThreadPoolInner` as the `Condvar` and `Mutex` pair as in the example you references. You should not need to create new condvar outside of `ThreadPoolInner::new()`.

> Can I join all workers in pool without for loop?

Yes, actually, this is done automatically, even if we leave the `drop` trait empty, as I mentioned in the above comment:

> Yes your understanding is correct. Actually for drop, even if you don't do anything, Rust will automatically drop all worker threads, (see [the reference](https://doc.rust-lang.org/reference/destructors.html#destructors))

`ThreadPoolInner` will call `drop` on `workers` `Vec`, which will call `drop` for each `Worker` in the `Vec`, which should call their `join`.

So that leaves your problem as either 1 or 2.

I think that your logic for implementing `wait_empty()` is correct, if you use the given condvar and mutex. Can you try fixing your `wait_empty()` first? If you can't seem to fix it on your own, can you file a new issue just for that, as it is a separate problem?

**Comment by deepkyu** [comment]
https://github.com/kaist-cp/cs431/issues/633#issuecomment-1253150752

This comment really helps a lot for the newbie like me :)
Thank you @Lee-Janggun !
